documentation updates for `LocalState` (#12781)
apollographql/apollo-client: docs/source/local-state/local-resolvers.mdx

1. `obj`: The object containing the result returned from the resolver on the parent field or the `ROOT_QUERY` object in the case of a top-level query or mutation.
2. `args`: An object containing all of the arguments passed into the field. For example, if you called a mutation with `updateNetworkStatus(isConnected: true)`, the `args` object would be `{ isConnected: true }`.
3. `context`: The context argument provides additional information without the possibility of name clashes. It takes the following shape:
    ```ts
    {
      // The request context. By default this is of type `DefaultContext`,
      // but can be changed if a `context` function is provided to LocalState
      requestContext: TContextValue,
      // The client instance making the request
      client: ApolloClient,
      // Whether the resolver is run as a result of gathering exported variables
      // or resolving the value as part of the result
      phase: "exports" | "resolve"
    }
    ```
    To access the cache, use `context.client.cache`.

> 4. `info`: Information about the execution state of the query. You will probably never have to use this one.

Reviewer: ```suggestion
4. `info`: Information about the execution state of the query. You will rarely need to use this value.
```

Figured we should avoid the word "never" as it makes it seem like we are telling you to stay away, so changing this to be a little softer 🙂

const localState = new LocalState<Resolvers>({
  resolvers: {
    // Your resolvers will be type-checked against the Resolvers type
  }
});
```

You may also pass a `ContextValue` generic used to ensure the `context` function returns the correct type. This type is inferred from your resolvers if not provided:

```ts
new LocalState<Resolvers, ContextValue>({
  // The return value of this function will be type-checked
  context: (options) => ({
    // ...
  }),
  resolvers: {
    // ...
  }
});
```

#### Generating types for local resolvers

When using TypeScript, you can generate resolver types for your local resolvers using the `@apollo/client-graphql-codegen` package. This package provides a specialized GraphQL Code Generator plugin that creates resolver types compatible with Apollo Client's `LocalState`.

##### Installation

Install the GraphQL Code Generator plugin for local state:

```bash
npm install -D @apollo/client-graphql-codegen
```

##### Configuration

Add the local-state plugin to your `codegen.ts` configuration. It's recommended to use a separate schema file containing only your local fields to avoid generating resolver types for your entire remote schema:

> ```ts title="codegen.ts"

Reviewer: Seems like there is bad spacing in general with this code block. Probably best to fix it in the editor for accuracy as some of my suggestions might not actually be correct.

import type { CodegenConfig } from '@graphql-codegen/cli';
import type { LocalStatePluginConfig } from '@apollo/client-graphql-codegen/local-state';

>   const config: CodegenConfig = {

Reviewer: ```suggestion
const config: CodegenConfig = {
```

  // ... your existing configuration
    generates: {
    // ... your existing generates
    './src/__generated__/local-resolvers.ts': {
      schema: ['./src/local-schema.graphql'],
      plugins: [
        'typescript',
        '@apollo/client-graphql-codegen/local-state',
        ],
        config: {
          // Ensures you return a `__typename` for any `@client` fields that
          // return object or array types

>           nonOptionalTypename: true,
>         // Required if your local schema extends existing schema types
>         baseTypesPath: './src/__generated__/graphql',
>           // If you provide a `context` function to customize the context value,
>         // provide the path or type here
>         contextType: './src/context#MyContextType',

Reviewer: ```suggestion
          nonOptionalTypename: true,
          // Required if your local schema extends existing schema types
          baseTypesPath: './src/__generated__/graphql',
          // If you provide a `context` function to customize the context value,
          // provide the path or type here
          contextType: './src/context#MyContextType',
```

        } satisfies LocalStatePluginConfig,
      },
    },
  };

export default config;
```

##### Creating your local schema

Create a GraphQL schema file that defines your local-only fields:

```graphql title="src/local-schema.graphql"
extend type Query {
  isLoggedIn: Boolean!
  cartItems: [String!]!
}

extend type Mutation {
  login(token: String!): Boolean!
  addToCart(item: String!): [String!]!
}

extend type Product {
  isInCart: Boolean!
}
```

##### Using generated types

After running the code generator, you can use the generated `Resolvers` type with your `LocalState` instance:

> ```ts

Reviewer: This code block could use another look for spacing.

// ... run the query using client.query, a <Query /> component, etc.
```

Here when the `GET_LAUNCH_DETAILS` query is executed, Apollo Client looks for a local resolver associated with the `isInCart` field. Since we've defined a local resolver for the `isInCart` field in the `LocalState` instance, it finds a resolver it can use. This resolver function is run, then the result is calculated and merged in with the rest of the query result (if a local resolver can't be found, Apollo Client will check the cache for a matching field - see [Local data query flow](#local-data-query-flow) for more details).

Setting resolvers through the `LocalState`'s `resolvers` option or through its `addResolvers` method adds resolvers to the internal resolver map (refer to the [Local resolvers](#local-resolvers) section for more details concerning the resolver map). In the above example we added a  `isInCart` resolver, for the `Launch` GraphQL object type, to the resolver map. Let's look at the `isInCart` resolver function more closely:

```js
  resolvers: {
    Launch: {
      isInCart: (launch, _args, { client }) => {

>         const { cartItems } = client.cache.readQuery({ query: GET_CART_ITEMS });

Author: Good call.

### Code splitting

> Depending on the complexity and size of your local resolvers, you might not always want to define them up front, when you create your initial `ApolloClient` instance. If you have local resolvers that are only needed in a specific part of your application, you can dynamically call `client.localState.addResolvers` to dynamically add new resolvers. This can be really useful when leveraging techniques like route based code-splitting.

Reviewer: ```suggestion
Depending on the complexity and size of your local resolvers, you might not always want to define them up front, when you create your initial `ApolloClient` instance. If you have local resolvers that are only needed in a specific part of your application, you can call `client.localState.addResolvers` to dynamically add new resolvers. This can be really useful when leveraging techniques like route based code-splitting.
```

The word "dynamically" was used twice and I think removing the first one makes more sense.

> Depending on the complexity and size of your local resolvers, you might not always want to define them up front, when you create your initial `ApolloClient` instance. If you have local resolvers that are only needed in a specific part of your application, you can dynamically call `client.localState.addResolvers` to dynamically add new resolvers. This can be really useful when leveraging techniques like route based code-splitting.

Reviewer: Does it make sense to use `client.localState.addResolvers` or `localState.addResolvers`? In one example, `localState` is pulled out to a separate variable. I'm fine with either, but figured I'd pose the question.
